GitHub action to setup poetry

DMRobertson opened the pull request:
In getting Sydent's dependencies to be managed by poetry I have composed a reusable action `setup-python-poetry` which sets up poetry for CI to use. Other parts of the WIP changes in Sydent make use of this action, but this is the "lowest" build block with no dependencies. Therefore I'm putting it up for review first.

I did look at prior art here, mainly [`snok/install-poetry`](https://github.com/snok/install-poetry). But that action required the user to manually setup caching; I wanted that to Just Work (TM). There's also a PR [for `actions/setup-python`](https://github.com/actions/setup-python/pull/281) which proposes to add support for caching poetry installations. However, I think that is concerned with caching the poetry-managed venv only; I also wanted to cache the installation of poetry itself.

## Explanation

To see this in action, visit https://github.com/matrix-org/sydent/actions/runs/1752681044, which is part of https://github.com/matrix-org/sydent/pull/488. That is a Sydent PR which tests three layers of GitHub actions config:
1. In the Sydent PR itself, there is a single "pipeline" workflow which uses [two callable workflows in "backend-meta"](https://github.com/matrix-org/sydent/blob/9a66fec9698f1ca5634f1e4b1f039a16cd885b42/.github/workflows/pipeline.yml#L24-L29) and [the `setup-python-poetry` action](https://github.com/matrix-org/sydent/blob/9a66fec9698f1ca5634f1e4b1f039a16cd885b42/.github/workflows/pipeline.yml#L39-L44).
2. My branch in `backend-meta` contains two ["reusable workflows"](https://docs.github.com/en/actions/using-workflows/reusing-workflows): one for [linting a poetry project](https://github.com/matrix-org/backend-meta/blob/dmr/poetry-ci/.github/workflows/python-ci.yml), and one for [building wheels and sdists](https://github.com/matrix-org/backend-meta/blob/dmr/poetry-ci/.github/workflows/packaging.yml). It's not necessary to put these in another repo, but I want to try and provide reusable building blocks for other projects. The linter workflow uses [the `setup-python-poetry` action](https://github.com/matrix-org/backend-meta/blob/af1ee9f43a5fd50e0bf6b52a02ecf50398cfde7f/.github/workflows/python-ci.yml#L15-L16) too.
3. `setup-python-poetry` is the action defined in this repo and proposed in the current PR. 

`setup-python-poetry` is a ["composite action"](https://docs.github.com/en/actions/creating-actions/creating-a-composite-action). It `pip`-installs poetry, then uses poetry to create a new virtual environment according to `pyproject.toml`. `actions/setup-python` does the hard work for us, but we make use of`actions/cache` to avoid having to re-install poetry and the project unless the python version, poetry version, or the lockfile has changed between CI runs.  

After using this action, subsequent job steps can use `poetry run foo [...]` to run `foo [...]` in the virtual environment.

If accepted, [best practice for actions](https://docs.github.com/en/actions/creating-actions/about-custom-actions#using-tags-for-release-management) says we should make a release branch and tag the appropriate version as `v1`.

Commit:
GitHub action to setup poetry

actions/setup-python does the hard work for us, but we make use of
actions/cache to avoid having to re-install poetry and the project
unless the poetry version we're using has changed, or the hash of
the lockfile has changed.

After using this action, subsequent job steps can use `poetry run foo
[...]` to run `foo [...]` in the virtual environment.

reivilibre commented on action.yml:
```diff
@@ -0,0 +1,83 @@
+      run: echo "::set-output name=dir::$(python -m poetry config virtualenvs.path)"
+      shell: bash
+
+    - name: Restore poetry venv
```
Calling this 'Restore' is a little bit confusing because I think this is also saving to the cache afterwards?

DMRobertson replied:
In one of the final steps, yes (but only if there wasn't a cache hit in the first place)

DMRobertson replied:
How about `Restore/cache foo bar`?

reivilibre commented on action.yml:
```diff
@@ -0,0 +1,83 @@
+      shell: bash
+
+    # For debugging---let's just check what we're working with.
+    - name: Dump virtual environment
+      if: "${{ inputs.debug == 'true' }}"
+      run: |
+        poetry env info
+        poetry show
+      shell: bash
```
If this output is only small I would be tempted to emit it unconditionally — when CI fails it's a nuisance to have to restart it all up in debug mode (?) to find out why, when it could just have been in the logs.

reivilibre commented on action.yml:
```diff
@@ -0,0 +1,83 @@
+      with:
+        python-version: ${{ inputs.python-version }}
+
+    - name: Check for poetry lock
+      run: "test -f poetry.lock || (echo No lock file! && false)"
+      shell: bash
```
Is the poetry lock file a necessary requirement? What about for libraries?

DMRobertson replied:
I'm proposing that we only use this action for projects with a lockfile.

The conventional wisdom for libraries is not to commit a lockfile. That means each CI run would end formally resolving dependencies, which can be sluggish. See for instance [this older attempt in signedjson](https://github.com/matrix-org/python-signedjson/runs/4952543111?check_suite_focus=true): `poetry install --no-interaction` took 22 seconds.

I suspect that building a wheel and installing it in editable mode with pip would be quicker. But more generally I've just kicked the "what do we do about libraries" can down the road.

reivilibre commented on action.yml:
```diff
@@ -0,0 +1,83 @@
+name: Setup Python and Poetry
+description: Setup Python and Poetry. Cribbed from snok/install-poetry.
```
Do we need to carry any licensing info across, and/or perhaps link to it in a hypothetical README?

DMRobertson replied:
"Cribbed" is maybe the wrong word. I looked to see how that action worked, but in the end decided I wanted different behaviour and wrote most of the yaml from scratch. snok's action was very useful for understanding what is/was going on with poetry though.

I'll acknowledge this and mark this as MIT licensed to match snok/install-poetry. I know we normally use Apache 2, but given that this is CI config rather than a core software component I think that's fine.

reivilibre reviewed (approved):
Seems reasonable.

Is having one repo per action the thing we're meant to do? It feels very heavyweight but I suppose you can't help it if you can't help it.

DMRobertson commented:
> Is having one repo per action the thing we're meant to do? It feels very heavyweight but I suppose you can't help it if you can't help it.

You can have actions kept within a project (they recommend a subdirectory of `.github/actions`) and refer to them by path. I tried putting this action into `backend-meta`, but when I called `backend-meta`'s reusable workflow from the sydent PR, the CI was looking for the action in the sydent repo, not the backend-meta repo. I tried a few other things and got syntax errors for my trouble. In the end I decided to stick with the thing that works, even if I didn't like it.

Commit:
Drop the debug input

Commit:
README

Commit:
LICENSE

Given that I've cited snok/install-poetry I've used theirs but stuck in
m.org as a secondary author

Commit:
Tweak name

Commit:
Small README writing tweaks
